Add and use stream metadata for RHCOS, add openshift-install coreos print-stream-json

cgwalters opened the pull request:
This implements part of the plan from:
https://github.com/openshift/os/issues/477

When we originally added the pinned RHCOS metadata `rhcos.json`
to the installer, we also changed the coreos-assembler `meta.json`
format into an arbitrary new format in the name of some cleanups.
In retrospect, this was a big mistake because we now have two
formats.

Then Fedora CoreOS appeared and added streams JSON as a public API.

We decided to unify on streams metadata; there's now a published
Go library for it: https://github.com/coreos/stream-metadata-go

Among other benefits, it is a single file that supports multiple
architectures.

UPI installs should now use stream metadata, particularly
to find public cloud images.

This is an important preparatory step for exposing this via
`oc` as well as having something in the cluster update to
it.

LorbusChris commented:
If we still have the OKD fork in 4.8, we'll want this there as well.
/cc @vrutkovs
/cc

cgwalters commented:
OK rebased :surfer: though only compile tested so far.  Let's run this for some platforms:
/test e2e-aws
/test e2e-azure
/test e2e-libvirt

cgwalters commented:
Picking this back up, but it will need https://github.com/coreos/coreos-assembler/pull/2052

cgwalters commented:
/test e2e-aws
/test e2e-azure
/test e2e-libvirt

cgwalters commented:
/test e2e-aws
/test e2e-azure
/test e2e-libvirt

cgwalters commented:
OK cool!  e2e-libvirt passed, and the other two just failed on some sort of Prometheus flake/issue that is very unlikely related to this.

cgwalters commented:
OK lifting draft!

staebler commented on data/data/rhcos-4.8.json:
```diff
@@ -0,0 +1,394 @@
+{
+  "stream": "rhcos-4.8",
```
What is the significance of the name of the stream? Does the file name have to match the name of the stream, or is that just convention?

cgwalters replied:
Convention, see e.g.
https://builds.coreos.fedoraproject.org/streams/stable.json
as another example.

Here we're matching the OpenShift versioning.

staebler replied:
My concern is the maintenance involved with code and docs if this file changes names with every release.

Even with this PR where the name change is fresh, there are a number of locations in this repo where the old rhcos.json name is still used.
```
$ grep -rnE . -e "rhcos\.json" --exclude-dir vendor --exclude-dir .git
./pkg/types/vsphere/platform.go:27:	// ClusterOSImage overrides the url provided in rhcos.json to download the RHCOS OVA
./images/installer/Dockerfile.upi.ci:17:COPY --from=builder /go/src/github.com/openshift/installer/data/data/rhcos-4.8.json /var/lib/openshift-install/rhcos.json
./data/data/install.openshift.io_installconfigs.yaml:1760:                    description: ClusterOSImage overrides the url provided in rhcos.json
./docs/user/azure/install_upi.md:201:export VHD_URL=`curl -s https://raw.githubusercontent.com/openshift/installer/release-4.3/data/data/rhcos.json | jq -r .azure.url`
./docs/user/azure/install_upi.md:207:export VHD_URL=`curl -s https://raw.githubusercontent.com/openshift/installer/master/data/data/rhcos.json | jq -r .azure.url`
./docs/user/metal/customization_ipi.md:28:[/data/data/rhcos.json](/data/data/rhcos.json), but they can be
./docs/user/metal/customization_ipi.md:32:uncompressed SHA256 from rhcos.json.
./docs/user/gcp/install_upi.md:473:export IMAGE_SOURCE=$(curl https://raw.githubusercontent.com/openshift/installer/master/data/data/rhcos.json | jq -r .gcp.url)
./docs/user/openstack/customization.md:115:Normally the installer downloads the RHCOS image from a predetermined location described in [data/data/rhcos.json](/data/data/rhcos.json)). But the download URL can be overridden, notably for disconnected installations.
```

staebler commented on go.mod:
```diff
@@ -21,6 +21,7 @@ require (
 	github.com/clarketm/json v1.14.1
 	github.com/containers/image v3.0.2+incompatible
 	github.com/coreos/ignition/v2 v2.9.0
+	github.com/coreos/stream-metadata-go v0.0.0-20210216154348-7c4d5d7d95b5
```
Please move vendor changes into their own commit.

staebler commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -21,6 +21,7 @@ import (
 	azureprovider "sigs.k8s.io/cluster-api-provider-azure/pkg/apis/azureprovider/v1beta1"
 	openstackprovider "sigs.k8s.io/cluster-api-provider-openstack/pkg/apis/openstackproviderconfig/v1alpha1"
 
+	coreosarch "github.com/coreos/stream-metadata-go/arch"
```
Move this import into the previous section.

staebler commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -338,16 +339,29 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
 		preexistingnetwork := installConfig.Config.GCP.Network != ""
 
-		imageRaw, err := rhcospkg.GCPRaw(ctx, installConfig.Config.ControlPlane.Architecture)
+		archName := coreosarch.RpmArch(fmt.Sprintf("%s", installConfig.Config.ControlPlane.Architecture))
```
Converting to a string works fine here.
```suggestion
		archName := coreosarch.RpmArch(string(installConfig.Config.ControlPlane.Architecture))
```

staebler commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -338,16 +339,29 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
+		st, err := rhcospkg.FetchCoreOSBuild(ctx)
 		if err != nil {
-			return errors.Wrap(err, "failed to find Raw GCP image URL")
+			return err
```
Wrap errors with meaningful messages.

cgwalters replied:
OK rather than doing this in all the callers I'm going to add some more error wrapping/prefixing in the functions; that OK by you?  I tried to also do that in stream-metadata-go, see e.g.

https://github.com/coreos/stream-metadata-go/blob/7c4d5d7d95b51e9fd2043375ee4b727f68c8f96a/stream/stream_utils.go#L15

staebler replied:
If there is enough meaningful context to track down the error, that I am satisfied, regardless of where that context is added.

staebler commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -338,16 +339,29 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
-			return errors.Wrap(err, "failed to find Raw GCP image URL")
+			return err
+		}
+		starch, err := st.GetArchitecture(archName)
```
Can we change the name of the `starch` variable so that it is not a real word that has nothing to do with what the variable represents?
```suggestion
		stArch, err := st.GetArchitecture(archName)
```
or
```suggestion
		streamArch, err := st.GetArchitecture(archName)
```

cgwalters replied:
Hah, honestly I hadn't even noticed it was a real world and I definitely speak the English.

staebler commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -338,16 +339,29 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
+			return err
+		}
+
+		var gcpimage string
+		if starch.Images.Gcp != nil {
+			img := starch.Images.Gcp
+			gcpimage = fmt.Sprintf("projects/%s/global/images/%s", img.Project, img.Name)
+		} else {
+			return fmt.Errorf("%s: No GCP build found", st.FormatPrefix(archName))
 		}
```
```suggestion
		img := starch.Images.Gcp
		if img == nil {
			return fmt.Errorf("%s: No GCP build found", st.FormatPrefix(archName))
		}
		gcpimage := fmt.Sprintf("projects/%s/global/images/%s", img.Project, img.Name)
```

staebler commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -338,16 +339,29 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
+			img := starch.Images.Gcp
+			gcpimage = fmt.Sprintf("projects/%s/global/images/%s", img.Project, img.Name)
+		} else {
+			return fmt.Errorf("%s: No GCP build found", st.FormatPrefix(archName))
```
This error wrapping, there will be a lot of colons. Let's not add to that.
```suggestion
			return fmt.Errorf("No GCP build found for %s architecture", st.FormatPrefix(archName))
```

cgwalters replied:
Already taken care of by the `FormatPrefix` call: https://github.com/coreos/stream-metadata-go/blob/7c4d5d7d95b51e9fd2043375ee4b727f68c8f96a/stream/stream_utils.go#L7

cgwalters replied:
Ah yes, it would show up as 
`stream:rhcos-4.8 arch:x86_64: No GCP build found`
which I agree is ugly.  Hmm.  Maybe I'll change `FormatPrefix` to do instead `rhcos-4.8/x86_64:` ?

cgwalters replied:
https://github.com/coreos/stream-metadata-go/pull/17

staebler replied:
My concern is more the final resulting message that is presented to the user, not the isolated error here. The final error is going to look something like the following.
```
FATAL failed to fetch something: failed to do something else: failed to do another thing: stream:rhcos-4.8 arch:x86_64: No GCP build found
```
We are already using colons to separate different layers of the final error. The colon to separate between parts of an individual layer conflicts with that.

cgwalters replied:
Yep agreed.    The above PR merged and I've updated the vendor commit to use it.

cgwalters replied:
I'd note in *this particular* case this error really can't happen because we're fetching this data from a JSON blob compiled into the same binary; we're never going to update the installer with an RHCOS build not in GCP.  But the general error handling philosphy definitely makes sense!

staebler commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -338,16 +339,29 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
 				MasterConfigs:      masterConfigs,
 				WorkerConfigs:      workerConfigs,
-				ImageURI:           imageRaw,
+				ImageURI:           gcpimage,
```
Is there any difference between this image URI and the one from the rhcos image asset any more? This code used to get the "raw image" but now gets the same image as the rhcos image asset does. Is that intentional?

cgwalters replied:
OK right, I should have looked at this more closely.  Some quick digging with git annotate points to
https://github.com/openshift/installer/commit/1f2d071f07ffd9e12fb9df64faadc3627db55bbd
which is related to the special licenses dance.  But in fact, we enable the nested virt license on the public RHCOS image by default since 4.6, which was the main motivation to support license overrides introduced in https://github.com/openshift/installer/pull/3430

So...we could do a prep commit which cleans this up and drops the support from the installer for overriding licenses?

cgwalters replied:
(Random aside, bits of CoreOS CI testing actually relies on a nested KVM setup in our own Prow instance)

staebler replied:
A prep commit would be nice so as not to lose the nuance of this change.

I am not very familiar with how the licenses work. Are you claiming that we do not need to support user-supplied licenses any more? If we switch away from using the raw image, we will lose support for user-supplied licenses?

cgwalters replied:
https://github.com/openshift/installer/pull/4696

staebler replied:
That is what I was afraid of. We cannot remove support for licenses. It is a breaking change.

cgwalters replied:
> Are you claiming that we do not need to support user-supplied licenses any more?

I am 93.2% sure there is no use case for other licenses beyond `enable-vmx` (which again is the default now), I looked and couldn't even find references to this.  I tried out the "license list" API linked there:
https://cloud.google.com/compute/docs/reference/rest/v1/licenses/list?apix_params=%7B%22project%22%3A%22openshift-gce-devel%22%2C%22returnPartialSuccess%22%3Atrue%7D
and I just got the empty set.

cgwalters replied:
OK I added a hack for this.  I'd still argue we can drop the license stuff, but if we decide it all needs to stay as is I can try plumbing through the image URL into the stream metadata.

cgwalters commented:
OK some bits resolved, leaving others open.

cgwalters commented:
/test e2e-libvirt
Flake.
/test e2e-aws-upgrade
Looks like existing bugs.

The `ci/prow/e2e-metal-ipi-ovn-ipv6` job looks like it's broken by the format change: `curl: (3) <url> malformed`

And so is `e2e-ovirt`:

```
Installing from release registry.build01.ci.openshift.org/ci-op-z2m7nqcm/release@sha256:9d8b91ca7c2d1b8a6f19d751c6586e42426d511e48d43e48d7b7bb60c84d7502
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xb8 pc=0x2e27901]
goroutine 1 [running]:
github.com/openshift/installer/pkg/asset/rhcos.osImage(0xc0003b0b40, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/openshift/installer/pkg/asset/rhcos/image.go:116 +0xd01 
```

Looking at those two.

cgwalters commented:
/test e2e-azure
Also we should get an Azure run in here.

cgwalters commented:
OK I think ovirt should be fixed now, and to fix all the UPI jobs my plan is to ratchet with https://github.com/openshift/release/pull/16355

cgwalters commented:
/test e2e-azure
/test e2e-metal
/test e2e-gcp

cgwalters commented:
/hold
I now created https://github.com/openshift/enhancements/pull/679 which details a plan that will get `openshift/installer` entirely out of the business of maintaining a JSON file with CoreOS metadata - the basic idea is we binary patch it in, the same way we do the release image.  Please comment there if the plan makes sense to you!

cgwalters commented:
/retest

cgwalters commented:
OK for now I went with the very dirty hack here https://github.com/openshift/enhancements/pull/679#issuecomment-796833422 - to repost, we just keep `rhcos.json` as it is and *add* stream JSON - IPI installs will use the latter.

I am hopeful that we could relatively quickly ratchet things like this:

- This PR merges
- ART adds automation to sync the stream JSON to https://mirror.openshift.com
- Our UPI CI jobs learn to fetch that
- Drop rhcos.json from here

But I think we'd still need an `openshift-install coreos print-stream-metadata` command or so for things like [baremetal](https://github.com/openshift-metal3/dev-scripts/blob/7e4800462fa7e71aaa9e4a7f4eb10166a6b1789c/rhcos.sh#L14).

staebler commented on pkg/rhcos/builds.go:
```diff
@@ -5,69 +5,66 @@ import (
 
-	return meta, nil
+// FindArtifactURL uses the first recognized artifact type; this
+// mainly abstracts over e.g. `qcow2.xz` and `qcow2.gz`.  The main
+// platform with multiple artifact types today is `metal`; don't
+// use this for that platform.
```
I don't understand the last sentence. The way that I read it is that this function is mainly to accommodate the `metal` platform, but also don't use this function for the `metal` platform.

staebler reviewed (commented):
This looks good to me. I have a couple questions and a couple nits.

So that I am not confusing things, am I correct that the old `rhcos.json` file is retained because there are CI jobs that do not yet understand the new stream format or are still relying on reading the rhcos.json file? And once the "port the UPI docs and CI jobs" is complete, we can delete the `rhcos.json` file, too? I assume that we will field bugs for a while afterwards from customer UPI processes that are reliant on `rhcos.json`.

cgwalters commented on images/baremetal/Dockerfile.ci:
```diff
@@ -12,6 +12,7 @@ RUN TAGS="libvirt baremetal" hack/build.sh
 FROM registry.ci.openshift.org/ocp/4.8:base
 COPY --from=builder /go/src/github.com/openshift/installer/bin/openshift-install /bin/openshift-install
 COPY --from=builder /go/src/github.com/openshift/installer/data/data/rhcos.json /var/cache/
+COPY --from=builder /go/src/github.com/openshift/installer/data/data/rhcos-stream.json /var/cache/
```
Since we're going to go with keeping the old format short term, and add a new CLI to output the data, I'll just drop this.

cgwalters commented:
OK addressed comments.

> So that I am not confusing things, am I correct that the old rhcos.json file is retained because there are CI jobs that do not yet understand the new stream format or are still relying on reading the rhcos.json file? And once the "port the UPI docs and CI jobs" is complete, we can delete the rhcos.json file, too? I assume that we will field bugs for a while afterwards from customer UPI processes that are reliant on rhcos.json.

Yep exactly...porting everything *before* landing this PR would be really hard.  If we can't remove the legacy `rhcos.json` for 4.8 it wouldn't be completely the end of the world, but I'm really hoping we can do so.

OK also, this adds a new `openshift-install coreos print-stream-json` command that actual UPI installs can use, and so can our UPI CI jobs.

cgwalters commented:
Also this adds the code to install the configmap with the stream metadata, which was the main original goal.

cgwalters commented:
OK I cleaned up the UPI docs to refer to a shared section on the new command to print the stream metadata.  And all this will need to be replicated in the enhancement *and* product docs, but I'll look at those after we've done any bikeshedding on naming etc. here.

staebler commented on pkg/asset/manifests/coreosbootimage.go:
```diff
@@ -0,0 +1,99 @@
+)
+
+// CoreOSBootimages generates the manifest with CoreOS stream metadata for the bootimages
+type CoreOSBootimages struct {
```
We do not want to inject the ConfigMap to manifests that the cluster-bootstrap will create. We want to add the ConfigMap to the resources that the CVO creates. This is needed because the ConfigMap needs to be updated when the OpenShift version is upgraded.

cgwalters replied:
Nice catch - a great example of why code review is important.

However...this is a really interesting sub-topic that gets into the larger still-unanswered question of how we actually *use* this data once it's in the cluster.  See this thread https://github.com/openshift/enhancements/pull/679#discussion_r592390250

I think I vaguely had in the back of my mind that we'd defer this problem, the initial goal here just being having the data in-cluster at all.   Manual testing of machineset updates could involve just editing the configmap.

To be clear, I think you're right, we should update it by default. 

Anyways since you want this bit as a separate PR we can take the details of this there.

staebler commented on pkg/rhcos/builds.go:
```diff
@@ -5,69 +5,84 @@ import (
+// mainly abstracts over e.g. `qcow2.xz` and `qcow2.gz`.  (FCOS uses
+// xz, RHCOS uses gzip right now)
+// Some platforms have multiple artifact types, e.g. `metal` has an ISO
+// as well as PXE files, etc.  Avoid using this for any
+// use this for that platform.
```
This last sentence does not make sense.

staebler reviewed (commented):
This looks close to ready to go with a few caveats.
1) Can you take out the part of putting the ConfigMap into the cluster? Let's tackle that as a separate PR so that we don't hold up this PR further.
2) Can you break out the new command into a separate commit?

staebler commented on hack/update-rhcos-bootimage.py:
```diff
@@ -1,4 +1,23 @@
+# 
+# The initial file data/data/rhcos-4.8 was generated this way:
+# 
+# $ plume cosa2stream --name rhcos-4.8 --distro rhcos  x86_64=48.83.202102230316-0 s390x=47.83.202102090311-0 ppc64le=47.83.202102091015-0 > data/data/rhcos-4.8.json
```
I am also still not sold on the value of including the OpenShift version in the name of the rhcos.json file. It seems unnecessary for a file that is ultimately just bindata in the binary and would require changes in at least 3 files for every new version.

Prashanth684 replied:
maybe also put in a link to https://github.com/coreos/coreos-assembler/pull/2052 as a better reference?

cgwalters commented:
> Can you take out the part of putting the ConfigMap into the cluster? Let's tackle that as a separate PR so that we don't hold up this PR further.

That's the main point of this effort, to be clear.  Although this snowballed into exposing a command (+ new standardized format) which we need for UPI, so we could certainly refocus on doing that first and the configmap second.  OK.

cgwalters commented on docs/user/openstack/customization.md:
```diff
@@ -112,13 +112,17 @@ sshKey: ssh-ed25519 AAAA...
 
-To do so and upload binary data from a custom location the user may set `clusterOSImage` parameter in the install config that points to that location, and then start the installation. In all other respects the process will be consistent with the default.
+If you want to download the image manually, see [CoreOS bootimages](../overview.md#coreos-bootimages) for more information
+about bootimages.  This is useful for e.g. performing a disconnected installation.  To do this,
```
Combined these to "This is useful, for example, to perform a disconnected installation."  (The "for, ... for" stutters I think)

staebler replied:
:+1: Even better.

staebler commented:
> > Can you take out the part of putting the ConfigMap into the cluster? Let's tackle that as a separate PR so that we don't hold up this PR further.
> 
> That's the main point of this effort, to be clear. Although this snowballed into exposing a command (+ new standardized format) which we need for UPI, so we could certainly refocus on doing that first and the configmap second. OK.

Yep, I get that. I am just trying to break this down into smaller discrete parts mostly so that I don't have to re-review everything every time there is a change.

cgwalters commented:
- all comments addressed
- dropped configmap injection for now
- split CLI addition to a new commit

cgwalters commented:
Moved the configmap code to https://github.com/openshift/installer/pull/4760

staebler reviewed (commented):
/lgtm
/approve

staebler commented:
/hold for some more tests
/test e2e-gcp
/test e2e-azure

Prashanth684 commented:
@walters just curious- how would one generate the `rhcos-stream.json` ?

staebler commented:
> @cgwalters just curious- how would one generate the `rhcos-stream.json` ?

Oops. I think we lost that documentation somewhere along the way.

cgwalters commented:
> Oops. I think we lost that documentation somewhere along the way.

Fixed, dug up the changed `hack/update-rhcos-bootimage.py` from b6489889

staebler commented on hack/update-rhcos-bootimage.py:
```diff
@@ -1,4 +1,23 @@
 #!/usr/bin/env python3
+# As of 4.8 we are aiming to switch to stream metadata:
```
This is not a great place for this documentation, given that this python script is no longer used. Can we add this to a new file under docs/dev instead?

cgwalters replied:
The script unfortunately may still be used until we can port CI and other users off `rhcos.json`.  So I was planning to keep the script until that happens, but we can migrate this stuff then to docs/dev after that?  Or I guess I could make the doc now and point from there to here too.

staebler replied:
Oh, I missed that the rest of this file was left untouched, since the file had been deleted in the last revision that I reviewed.

cgwalters replied:
OK added a new docs/dev/pinned-coreos.md!

staebler commented on hack/update-rhcos-bootimage.py:
```diff
@@ -1,4 +1,23 @@
+# That transition hasn't yet fully completed; there are two copies of the
+# RHCOS metadata:
+# 
+# - data/data/rhcos-4.8.json (stream format, 4.8+)
```
This is not the correct filename any more.

staebler commented on docs/dev/pinned-coreos.md:
````diff
@@ -0,0 +1,44 @@
+To update the bootimage for one or more architectures, use e.g.
+
+```
+$ plume cosa2stream --target data/data/rhcos-4.8.json --distro rhcos  x86_64=48.83.202102230316-0 s390x=47.83.202102090311-0 ppc64le=47.83.202102091015-0
````
```suggestion
$ plume cosa2stream --target data/data/rhcos-stream.json --distro rhcos  x86_64=48.83.202102230316-0 s390x=47.83.202102090311-0 ppc64le=47.83.202102091015-0
```

staebler reviewed (approved):
/lgtm

cgwalters commented:
From [the openstack logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_installer/4582/pull-ci-openshift-installer-master-e2e-openstack/1374377942871707648/artifacts/e2e-openstack/bootstrap/nova.log)
> [ 1385.637740] ignition[718]: GET error: Get "https://image.rdo.mtl2.vexxhost.net//v2/images/d4862180-1038-4f8d-a447-03b5864793b9/file": dial tcp: lookup image.rdo.mtl2.vexxhost.net on 10.0.0.10:53: server misbehaving

Ah yeah, unrelated; https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/4772/pull-ci-openshift-installer-master-e2e-openstack/1372760778477670400 hit the same thing.

staebler commented:
Who wants to win the race between this PR and #4778?

cgwalters commented:
/test e2e-metal

cgwalters commented:
/test e2e-azure
/test e2e-gcp

cgwalters commented:
```
 Downloading https://files.pythonhosted.org/packages/53/7f/55721ad0501a9076dbc354cc8c63ffc2d6f1ef360f49ad0fbcce19d68538/pip-20.3.4.tar.gz#sha256=6773934e5f5fc3eaa8c5a44949b5b924fc122daa0a8aa9f80c835b4ca2a543fc
error: https://files.pythonhosted.org/packages/53/7f/55721ad0501a9076dbc354cc8c63ffc2d6f1ef360f49ad0fbcce19d68538/pip-20.3.4.tar.gz#sha256=6773934e5f5fc3eaa8c5a44949b5b924fc122daa0a8aa9f80c835b4ca2a543fc hostname 'files.pythonhosted.org' doesn't match either of 'r.ssl.fastly.net', '*.catchpoint.com', '*.cnn.io', '*.dollarshaveclub.com', '*.eater.com', '*.fastly.picmonkey.com', '*.files.saymedia-content.com', '*.ft.com', '*.meetupstatic.com', '*.nfl.com', '*.pagar.me', '*.picmonkey.com', '*.realself.com', '*.sbnation.com', '*.shakr.com', '*.streamable.com', '*.surfly.com', '*.theverge.com', '*.thrillist.com', '*.vox-cdn.com', '*.vox.com', '*.voxmedia.com', 'eater.com', 'ft.com', 'i.gse.io', 'picmonkey.com', 'realself.com', 'static.wixstatic.com', 'streamable.com', 'surfly.com', 'theverge.com', 'vox-cdn.com', 'vox.com', 'www.joyent.com'
error: build error: error building at STEP "RUN easy_install 'pip<21'": error while running runtime: exit status 1 
```
err...

staebler commented:
> ```
>  Downloading https://files.pythonhosted.org/packages/53/7f/55721ad0501a9076dbc354cc8c63ffc2d6f1ef360f49ad0fbcce19d68538/pip-20.3.4.tar.gz#sha256=6773934e5f5fc3eaa8c5a44949b5b924fc122daa0a8aa9f80c835b4ca2a543fc
> error: https://files.pythonhosted.org/packages/53/7f/55721ad0501a9076dbc354cc8c63ffc2d6f1ef360f49ad0fbcce19d68538/pip-20.3.4.tar.gz#sha256=6773934e5f5fc3eaa8c5a44949b5b924fc122daa0a8aa9f80c835b4ca2a543fc hostname 'files.pythonhosted.org' doesn't match either of 'r.ssl.fastly.net', '*.catchpoint.com', '*.cnn.io', '*.dollarshaveclub.com', '*.eater.com', '*.fastly.picmonkey.com', '*.files.saymedia-content.com', '*.ft.com', '*.meetupstatic.com', '*.nfl.com', '*.pagar.me', '*.picmonkey.com', '*.realself.com', '*.sbnation.com', '*.shakr.com', '*.streamable.com', '*.surfly.com', '*.theverge.com', '*.thrillist.com', '*.vox-cdn.com', '*.vox.com', '*.voxmedia.com', 'eater.com', 'ft.com', 'i.gse.io', 'picmonkey.com', 'realself.com', 'static.wixstatic.com', 'streamable.com', 'surfly.com', 'theverge.com', 'vox-cdn.com', 'vox.com', 'www.joyent.com'
> error: build error: error building at STEP "RUN easy_install 'pip<21'": error while running runtime: exit status 1 
> ```
> 
> err...

That is a transient error (at least it looks that way and I sure hope so).

cgwalters commented:
> That is a transient error (at least it looks that way and I sure hope so).

Hmm, maybe yes.  Best I can turn up is https://stackoverflow.com/questions/25981703/pip-install-fails-with-connection-error-ssl-certificate-verify-failed-certi
which seems like it might be that the `easy_install` command is hardcoding a list of domains, and this might be fixed in later versions except we're using an old rhel7 version (presumably that could be bumped to rhel8?).

staebler commented:
I am removing the hold so that we get automatic retests. The merge will be blocked by the infinitesimal pass rate of e2e-aws-upgrade anyway.
/hold cancel

cgwalters commented:
I don't think we get retests as long as the hold is up, OK to lift that?

cgwalters commented:
/test e2e-azure
/test e2e-gcp

cgwalters commented:
Alternatively https://github.com/openshift/installer/pull/4784 I think is working, so we can land that first.

staebler commented:
> I don't think we get retests as long as the hold is up, OK to lift that?

I already lifted the hold. https://github.com/openshift/installer/pull/4582#issuecomment-805141797

cgwalters commented:
/test e2e-azure
/test e2e-gcp
/test e2e-aws-upgrade

cgwalters commented:
/test e2e-metal

vrutkovs commented:
/refresh

vrutkovs commented:
/retest

vrutkovs commented:
/retest

cgwalters commented:
Now merges the changes from https://github.com/openshift/installer/pull/4778

vrutkovs commented:
/retest

staebler commented:
I assume that this is going to need a rebase now that #4778 has merged.

cgwalters commented:
Rebased :surfer:

staebler reviewed (commented):
/lgtm

staebler commented:
/override ci/prow/e2e-aws-upgrade

romfreiman commented:
🤞

romfreiman commented:
/retest
